### PR TITLE
feat(engine): Add formation pattern code and unit-tests

### PR DESCRIPTION
**Pull request by petervdmeer**
**Feature:** This PR implements just the formations pattern calculation from PR #4471

## Feature Details
This PR adds the formation-pattern loading and calculation code from #4471 and related unit-tests. I believe that this part of the formations-code is most mature (reviewed in the RFC, unit-tested, code-reviewed), while other parts of #4471 might be less "production-ready".

Note that merging this PR would not make the formations available to players of the game, it would only add the pattern-support for use by follow-up code-PRs. Merging this PR would however allow for independent development of other parts of #4471 (and should allow us to get the parts of #4471 that we like to get into the game, while keeping out the parts from #4471 that we don't want to have).

## UI Screenshots
See #4471.

## Usage Examples
See #4471.

## Testing Done
Unit tests (some more than in #4471) and see #4471 for all user-testing that was done.

## Performance Impact
None; this PR only adds the formation pattern calculation code and formation loading code, not the (real-time) code to actually use the formations.

### CODE REVIEW

**Comment by tehhowch** [comment]
The part I'm most tentative about with the formations stuff _is_ the user-facing syntax, so splitting apart the feature doesn't really get it closer to being mergable.

(It does help, yes, to separate the definitions from additional behavior like hot-swappable escort formations and hotkeys.)

**Comment by Zitchas** [comment]
We have had a bit of a discussion regarding the user-facing syntax in the past, and honestly I'm not sure how much better it can be made. No-one is going to be making formations without a pretty solid grasp of some combination of mathematics, co-ordinate geometry, trignometry, and/or some programmingesque pattern development. I doubt we'll ever have something where a random person can meaningfully add new formations. I'm not entirely sure we'll ever be able to make it meaningfully simpler, either. So I *really* don't think some vague desire to simplify it should be holding up a rather valuable and interesting mechanic.

That being said, the set of formations that petervdmeer has in the main/full PR is quite comprehensive. In the more-than-a-year that we have been poking at this, I still haven't come up with any new formations I'd like to see.

**Comment by petervdmeer** [comment]
@tehhowch The goal of splitting was indeed to focus on the part which is really blocking the merge; the user-facing syntax.
(Distractions like hotkeys, const-functions, position-swapping, and massive merge-conflicts should not affect this PR.)

I just updated the [RFC/specification](https://github.com/EndlessSkyCommunity/endless-sky/pull/52) to allow content creators to create formations:
- without complex mathematics (just basic "pixel-sized" coordinates relative to the formation leader).
- with mixed categories. (Which I remember you suggested some time ago.)
- and with somewhat improved keywords (position sounds better than slot or point).

This should allow formation pattern definitions like the following:

```
formation "Interceptor Escorted Convoy"
	position -600 600
		category "Interceptor"
	position 600 600
		category "Interceptor"
	position 600 0
		category "Interceptor"
	position -600 0
		category "Interceptor"
	position -600 -600
		category "Interceptor"
	position 600 -600
		category "Interceptor"
	position 500 0
		category "Transport" "Light Freighter" "Heavy Freighter"
	position -500 0
		category "Transport" "Light Freighter" "Heavy Freighter"
```
(This is a formation with Interceptors on each side and some transports in the middle)

I also don't see how we can make this better; the absolute minimum that one needs when defining formations is the positions where ships should appear, and the category is something which I remember you explicitly requested to allow a single formation definition to describe a complete arrangement of different types of ships.

If you have reservations to the format as described, then please provide concrete feedback on how to get to a mergable user-facing syntax.
And if you are okay with the format as described here, then please also let us know so that I can process the latest specification update into this PR.

**Comment by petervdmeer** [comment]
I processed the latest specification update into this PR.

**Comment by Zitchas** [comment]
If we're good with the mechanics, and simply have issues with the user-facing code, is there any reason we couldn't merge this so we can focus on resolving the objections with the user-facing stuff?

**Comment by petervdmeer** [comment]
> If we're good with the mechanics, and simply have issues with the user-facing code, is there any reason we couldn't merge this so we can focus on resolving the objections with the user-facing stuff?

This PR (intentionally) only contains the `user-facing syntax`. Misalignment on the `user-facing syntax` between @tehhowch and me is the only big issue blocking formation flying. Any other issue we encounter in follow-up PRs will be minor compared to this.

I feel that if tehhowch and me would meet each-other in person (near a whiteboard where we could draw on), then we would resolve this misalignment in minutes.
But I guess we need to do this the hard way; what I suggest is that I change this PR to the absolute minimum subset of [the RFC](https://github.com/EndlessSkyCommunity/endless-sky/pull/52) and [the initial input from tehhowch](https://github.com/endless-sky/endless-sky/pull/4418#pullrequestreview-258598505).

That means that the only `user-facing definition syntax` that we support with this PR will be:
```
formation <name>
	position <x> <y>
	position <x2> <y2>
```
(where the x and y coordinates are in pixels and we don't support anything more fancy than that)

This absolute minimum subset should allow us to get the foundation for formations into the game without alignment issues between tehhowch and me. I'll try to find some time to get the rework done that is needed on this PR to reduce it to only the subset.

**Comment by petervdmeer** [comment]
I reduced the user-facing syntax in this PR to only support this syntax:
```
formation <name>
	position <x> <y>
	position <x2> <y2>
```
If there are no objections to this syntax, then I suggest we follow @Zitchas suggestion and merge this PR.

**Comment by Hurleveur** [comment]
I don't see any problem with the syntax, personally, not any user can create missions with good text and all, this just takes another kind of skill to do. I think that in time smbd will develop a website or some other kind of tool to do these formations easily, and in the meantime its just additional options for some at least.

However, as for ships, missions, etc, users should still be able to use it, so if we have predefined formations ready to use that the player can swap to in game by clicking on a button I can't see any reason against formations.

**Comment by Zitchas** [comment]
Especially since someone already *did* create a website to test formations. It was just a demo (I think maybe for a school project) and it broke after an update to the formations stuff, if I recall correctly; and wasn't maintained since the formations stuff is still up in the air. But if we solidified on it I'm sure the webcode could be updated easily enough. We could even migrate it over to live adjacent to the ship hardpoint webpage.

File: source/FormationPattern.cpp
Comment (tibetiroka) [inline source/FormationPattern.cpp:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```
The standard copyright header was changed recently.

File: source/FormationPattern.h
Comment (tibetiroka) [inline source/FormationPattern.h:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

File: tests/unit/src/test_formationPattern.cpp
Comment (tibetiroka) [inline tests/unit/src/test_formationPattern.cpp:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

**Review by tibetiroka — CHANGES_REQUESTED** [review]

File: source/FormationPattern.cpp
Comment (warp-core) [inline source/FormationPattern.cpp:99]: Any reason not to pass these values to `emplace_back`?

File: source/FormationPattern.cpp
Comment (warp-core) [inline source/FormationPattern.cpp:123]: This comment doesn't seem to match what this method does.

File: source/FormationPattern.h
Comment (warp-core) [inline source/FormationPattern.h:85]: Any reason this doesn't return a reference?

File: source/GameData.cpp
Comment (warp-core) [inline source/GameData.cpp:513]: This should be below `Fleets()`.

**Review by warp-core — COMMENTED** [review]
While this type of definition certainly seems simple, it is also incredibly limiting, and can make even simple arrangements rather long and repetitive to define, which makes errors in definitions more difficult to find.
Even if we continue with this implementation for now, I think the more complex option should still be supported, even if that support is added at a later date.

**Review by tibetiroka — CHANGES_REQUESTED** [review]
Warp caught the other formatting issues.

File: source/FormationPattern.cpp
Comment (petervdmeer) [reply source/FormationPattern.cpp:83]: That is true, this version of the code doesn't allow overriding an existing formation.

If we find good reasons to allow such overrides this in the future (instead of only defining new formations in plugins), then this should be easy to change; just change this code to a warning and execute a delete of the existing formation before loading the override.

File: source/FormationPattern.cpp
Comment (petervdmeer) [reply source/FormationPattern.cpp:123]: Indeed, the comment belongs to a much more complex version of the formations code. I'll update it.

File: source/FormationPattern.cpp
Comment (petervdmeer) [reply source/FormationPattern.cpp:99]: Sounds like a good idea. I'll look into it.

File: source/FormationPattern.h
Comment (petervdmeer) [reply source/FormationPattern.h:25]: Body is indeed no longer used in this version of the code.

File: source/FormationPattern.h
Comment (petervdmeer) [reply source/FormationPattern.h:85]: Historical reasons; the previous more complex versions of this code calculated the positions on the fly, so there were no references to return.

The current version of the code only works with positions/Points, so we could indeed return references here.

Comment (petervdmeer) [reply source/FormationPattern.h]: I updated the code to completely remove this function. It now uses direct references to the positions in the formation, or to the default location (Point(0, 0)) when no valid position is available.

**Review by Zitchas — APPROVED** [review]
I tested #4471 extensively, and merged it with many other changes into Revamp, and it worked beautifully. As such, I consider my approval of that original PR to carry over to here.

Note: this is purely an approval based desirability, on testing and gameplay, and content creator use in game content. I cannot vouch for the code itself.

File: source/FormationPattern.cpp
Comment (Hurleveur) [inline source/FormationPattern.cpp:23]: Dont we usuallly have only 1 space between this and the namespace after?

File: source/FormationPattern.cpp
Comment (Hurleveur) [inline source/FormationPattern.cpp:53]: There is smth weird about this code to me; there is no way to go back to the start of the formation?

File: tests/unit/src/test_formationPattern.cpp
Comment (Hurleveur) [inline tests/unit/src/test_formationPattern.cpp:28]: Not sure if this is the standard amount of newlines here (and on phone so cant really check)

**Review by Hurleveur — CHANGES_REQUESTED** [review]
I like this code. Something troubles me though, there is no looping possible currently, no way to restart the formation to its beginning.
As I said before I support having 2 syntaxs; this more simple one for a small fleet, and the more complex ones for a bigger one. As ES generally is about small fleets, this way feels the most useful to start with.

It isnt actually used/usable anywhere in the game currently, its just for testing purposes. So it cant do any harm I guess but at the same time its not really a feature.

File: source/FormationPattern.cpp
Comment (petervdmeer) [reply source/FormationPattern.cpp:53]: It is an iterator. If you want to restart, then throw away the current iterator and create a new one with FormationPattern::begin().

We could extend the iterator with a reset function if we like, but it is not strictly required.

File: tests/unit/src/test_formationPattern.cpp
Comment (petervdmeer) [reply tests/unit/src/test_formationPattern.cpp:28]: The other tests only seem to have 1 newline here. I'll remove one.

File: source/FormationPattern.cpp
Comment (petervdmeer) [reply source/FormationPattern.cpp:23]: Yes, looks like I added 2 newlines too much. I'll remove them.

File: source/FormationPattern.cpp
Comment (Hurleveur) [reply source/FormationPattern.cpp:51]: Iterators go ++ and -- from my understanding of it but as this has no practical applications yet it doesnt matter much I guess?

Comment (petervdmeer) [reply source/FormationPattern.cpp:51]: Most iterators are "Random Access" or "Bidirectional" and indeed go in 2 directions. But i also found single direction input iterators in [a nice overview table](https://cplusplus.com/reference/iterator/) and those input iterators seemed sufficient for calculating the sequence of positions that was required.

When implementing other formations code, then we might need to implement something more complex (depending on other design choices that we make), but for now this code seems to suffice for testing of the positions code (and that is all we do in this PR).

**Review by Hurleveur — APPROVED** [review]
I wish it wouldnt show me approving as the team but just as part of it. Anyway, I'm good with it being this way, even if its a bit incomplete, there is no reason to not allow this into the game; it allows us to build more on it later on without changing anything yet, and sets good bases of testing.

**Comment by warp-core** [comment]
> I wish it wouldnt show me approving as the team but just as part of it.

Dismiss the request before submitting your review.

**Comment by petervdmeer** [comment]
This PR is code-reviewed, and only has the absolute minimum subset of position-specification functionality which everyone seems to agree on.
Discussions on the dynamic, mathematical complex, position calculations no longer apply to this PR, since those complex calculations were taken out in the latest version.
Let's merge and focus on getting a simple first version of formation flying (without the complex mathematics) available in ES.

**Pull request merged**
